assembly/src/assembler/mod.rs

```diff
@@ -695,3 +712,65 @@ fn combine_mast_node_ids(
 
     Ok(mast_node_ids.remove(0))
 }
+
+fn merge_contiguous_basic_blocks(
+    mast_node_ids: Vec<MastNodeId>,
+    mast_forest_builder: &mut MastForestBuilder,
+) -> Result<Vec<MastNodeId>, AssemblyError> {
+    let mut merged_mast_node_ids = Vec::with_capacity(mast_node_ids.len());
+    let mut contiguous_basic_block_ids: Vec<MastNodeId> = Vec::new();
+
+    for mast_node_id in mast_node_ids {
+        if mast_forest_builder[mast_node_id].is_basic_block() {
+            contiguous_basic_block_ids.push(mast_node_id);
+        } else {
+            if let Some(merged_basic_block_id) =
+                merge_basic_blocks(&contiguous_basic_block_ids, mast_forest_builder)?
+            {
+                merged_mast_node_ids.push(merged_basic_block_id)
+            }
+            contiguous_basic_block_ids.clear();
+
+            merged_mast_node_ids.push(mast_node_id);
+        }
+    }
+
+    if let Some(merged_basic_block_id) =
+        merge_basic_blocks(&contiguous_basic_block_ids, mast_forest_builder)?
+    {
+        merged_mast_node_ids.push(merged_basic_block_id)
+    }
+
+    Ok(merged_mast_node_ids)
+}
+
+fn merge_basic_blocks(
+    contiguous_basic_block_ids: &[MastNodeId],
+    mast_forest_builder: &mut MastForestBuilder,
+) -> Result<Option<MastNodeId>, AssemblyError> {
```

Comment:
done

Reply:
Ad discussed offline, we could probably move the whole `join_mast_node_ids()` into the `MastForestBuilder`. I'd probably name it just `join_nodes()`:
```rust
impl MastForestBuilder {

    pub fn join_nodes(&mut self, node_ids: Vec<MastNodeId>) -> Result<MastNodeId, AssemblyError> {

    }

}
```
Then, in this function we'd keep track of eliminated nodes and save them into an instance variable of the `MastForestBuilder`.

Reply:
Done. What I noticed while implementing is that when we merge a basic block, we can't remove it yet. We first need to check that no other block uses it before we can remove it. This is because the `MastForestBuilder` will return an existing `MastNodeId` when trying to add a node that hashes to something already in the forest.

Specifically [this test line](https://github.com/0xPolygonMiden/miden-vm/blob/2e4b88947e4d6a4f100480972772a49767501e49/assembly/src/tests.rs#L1271) shows this case: `mul` is used in a basic block that gets merged, but is also reused by the `if.true mul end`. Hence, we cannot remove the `mul` basic block.